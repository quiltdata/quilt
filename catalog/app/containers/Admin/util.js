import mapValues from 'lodash/mapValues';
import omit from 'lodash/omit';
import { compose, withHandlers, withProps, withStateHandlers } from 'recompose';

export const formatDate = (d) => d ? new Date(d).toLocaleString() : 'N/A';

export const withStatefulActions = (bindActions) => compose(
  withStateHandlers({ pending: {} }, {
    startAction: ({ pending }) => (key) => ({
      pending: Object.assign({}, pending, { [key]: true }),
    }),
    endAction: ({ pending }) => (key) => ({
      pending: omit(pending, [key]),
    }),
  }),
  withHandlers(({ actions }) =>
    mapValues(actions, (action, key) =>
      // eslint-disable-next-line object-curly-newline
      ({ pending, startAction, endAction, actions: { [key]: callback } }) => (arg) => {
        if (pending[arg]) return;
        startAction(arg);
        callback(arg)
          .then(() => endAction(arg))
          .catch((e) => { endAction(arg); throw e; });
      })),
  withProps((props) => ({ bindActions: bindActions(props) })),
);
