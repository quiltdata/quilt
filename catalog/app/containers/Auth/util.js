import { stringify } from 'query-string';

import { LATENCY_SECONDS } from './constants';

export const makeHeadersFromTokens = ({ token }) => ({
  Authorization: `Bearer ${token}`,
});

const makeURL = (url, query) => {
  const qs = stringify(query);
  const search = qs ? `?${qs}` : '';
  return `${url}${search}`;
};

const getPath = (loc) => loc.pathname + loc.search;

export const makeSignInURL = (next = getPath(window.location)) =>
  // TODO(dima): Sign up vs sign in?
  makeURL('/signin', { next });

export const adjustTokensForLatency = (tokens) => ({
  ...tokens,
  expires_at:
    Number.isFinite(tokens.expires_at)
      ? tokens.expires_at - LATENCY_SECONDS
      : tokens.expires_at,
});
