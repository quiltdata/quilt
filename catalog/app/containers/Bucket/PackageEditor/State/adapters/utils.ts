import type { FileWithPath } from 'react-dropzone'

import { Status } from 'components/FileManager/FileRow'
import type { TreeEntry } from 'components/FileManager/FileTree'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export function calcChildren(
  entry: Model.PackageEntry | Model.S3File | FileWithPath,
  tailParts: string[],
  children: TreeEntry[] = [],
): TreeEntry[] {
  const [name, ...tail] = tailParts
  const found = children.find((child) => child.id === name)
  if (found) {
    if (tail.length) {
      found.children = calcChildren(entry, tail, found.children)
    }
    return children
  }

  return children.concat({
    id: name,
    name: tail.length ? s3paths.ensureSlash(name) : name,
    size: tail.length ? 0 : entry.size,
    status: Status.Unchanged,
    children: tail.length ? calcChildren(entry, tail, []) : undefined,
  })
}

export function sortEntries(entries: TreeEntry[]): TreeEntry[] {
  return [...entries].sort((a, b) => {
    if (a.children && !b.children) return -1
    if (!a.children && b.children) return 1
    return a.name.localeCompare(b.name)
  })
}
