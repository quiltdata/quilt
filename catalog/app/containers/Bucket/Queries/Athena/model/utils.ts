export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no data. It is not initialized
// `Loading` is loading
// `Error` is error
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

export function wrapData<T>(data: Data<T>, setPrev: (d: T) => void): DataController<T> {
  return {
    data,
    loadMore: () => hasData(data) && setPrev(data),
  }
}

export interface List<T> {
  list: T[]
  next?: string
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not initialized
// `Loading` is loading
// `Error` is error
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

export function wrapValue<T>(
  value: Value<T>,
  setValue: (d: T | null) => void,
): ValueController<T> {
  return {
    value,
    setValue,
  }
}

/** Data is loaded, or value is set to actual value */
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

/** No value yet: value or data was just initialized */
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

/** Data is loading, or value is waiting for data */
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

/** Value is selected with some or no value, or resolved with error, or data is loaded (successfully or not) */
export function isReady<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

/** Value is selected with some or no value, or data is loaded successfully */
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

/** User explicitly set no value */
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
