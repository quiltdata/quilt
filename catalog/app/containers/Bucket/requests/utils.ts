import * as R from 'ramda'

export const decodeS3Key = R.pipe(R.replace(/\+/g, ' '), decodeURIComponent)
