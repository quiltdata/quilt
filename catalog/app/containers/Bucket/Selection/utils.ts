import { join } from 'path'

import * as R from 'ramda'

import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export interface SelectionItem {
  logicalKey: string
}

export interface ListingSelection {
  [prefixUrl: string]: SelectionItem[]
}

export const EMPTY_MAP: ListingSelection = {}

interface SelectionHandles {
  [prefixUrl: string]: Model.S3.S3ObjectLocation[]
}

const convertIdToHandle = (
  id: string | number,
  parentHandle: Model.S3.S3ObjectLocation,
): Model.S3.S3ObjectLocation => ({
  bucket: parentHandle.bucket,
  key: join(parentHandle.key, id.toString()),
})

export const toHandlesMap = (selection: ListingSelection): SelectionHandles =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => ({
      ...memo,
      [prefixUrl]: items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    }),
    {} as SelectionHandles,
  )

export const toHandlesList = (selection: ListingSelection): Model.S3.S3ObjectLocation[] =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => [
      ...memo,
      ...items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    ],
    [] as Model.S3.S3ObjectLocation[],
  )

const mergeWithFiltered =
  (prefix: string, filteredItems: SelectionItem[]) => (allItems: SelectionItem[]) => {
    if (!allItems || !allItems.length) return filteredItems
    const selectionOutsideFilter = allItems.filter(
      (item) => !item.logicalKey.startsWith(prefix),
    )
    const newIds = [...selectionOutsideFilter, ...filteredItems]
    return R.equals(newIds, allItems) ? allItems : newIds // avoids cyclic update
  }

export function merge(
  items: SelectionItem[],
  bucket: string,
  path: string,
  filter?: string,
): (state: ListingSelection) => ListingSelection {
  const prefixUrl = `s3://${bucket}/${path}`
  const lens = R.lensProp<Record<string, SelectionItem[]>>(prefixUrl)
  return filter ? R.over(lens, mergeWithFiltered(filter, items)) : R.set(lens, items)
}

const EmptyKeys: string[] = []

export const getDirectorySelection = (
  selection: ListingSelection,
  bucket: string,
  path: string,
) => selection[`s3://${bucket}/${path}`] || EmptyKeys
