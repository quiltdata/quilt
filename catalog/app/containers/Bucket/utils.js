import * as React from 'react'

import * as AWS from 'utils/AWS'

export const withDownloadUrl = (handle, callback) => (
  <AWS.Signer.Inject>
    {(signer) =>
      callback(
        signer.getSignedS3URL(handle, { ResponseContentDisposition: 'attachment' }),
      )
    }
  </AWS.Signer.Inject>
)
