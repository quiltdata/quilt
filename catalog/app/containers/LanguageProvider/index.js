/*
 *
 * LanguageProvider
 *
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { mapProps, setPropTypes } from 'recompose';
import { createSelector } from 'reselect';
import { IntlProvider } from 'react-intl';

import { composeComponent } from 'utils/reactTools';
import { injectReducer } from 'utils/ReducerInjector';

import { REDUX_KEY } from './constants';
import reducer from './reducer';
import { makeSelectLocale } from './selectors';

export default composeComponent('LanguageProvider',
  injectReducer(REDUX_KEY, reducer),
  connect(createSelector(
    makeSelectLocale(),
    (locale) => ({ locale })
  )),
  setPropTypes({
    locale: PropTypes.string,
    messages: PropTypes.object,
    children: PropTypes.element.isRequired,
  }),
  mapProps(({ locale, messages, ...props }) => ({
    locale,
    key: locale,
    messages: messages[locale],
    ...props,
  })),
  IntlProvider);
