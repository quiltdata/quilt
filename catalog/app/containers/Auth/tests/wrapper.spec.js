import createHistory from 'history/createMemoryHistory';
import { fromJS } from 'immutable';
import React from 'react';
import { LOCATION_CHANGE } from 'react-router-redux';

import LanguageProvider from 'containers/LanguageProvider';
import { translationMessages as messages } from 'i18n';
import { Provider as APIProvider } from 'utils/APIConnector';
import { nest } from 'utils/reactTools';
import RouterProvider from 'utils/router';
import StoreProvider from 'utils/StoreProvider';
import { timestamp } from 'utils/time';
import configureStore from 'store';

import feature from 'testing/feature';
import errorSteps from 'testing/error';
import reactSteps from 'testing/react';
import requestsSteps from 'testing/requests';
import storageSteps from 'testing/storage';
import { mockComponentSelector } from 'testing/util';

import {
  Provider as AuthProvider,
  requireAuth,
  errors,
  apiMiddleware,
} from '..';

import { api, date, tokensRaw, datasets } from './fixtures';

jest.mock('material-ui/RaisedButton');
jest.mock('react-router-dom');
jest.mock('components/Error');
jest.mock('components/Working');
jest.mock('constants/config');
jest.mock('utils/time');
jest.mock('utils/errorReporting');

const Home = () => <h1>home</h1>;
const ProtectedHome = requireAuth(Home);

const homePath = '/home?q=sup';

const requests = {
  refreshTokens: {
    setup: () => ['post', '/refresh'],
    success: () => tokensRaw,
  },
};

const screens = {
  placeholder: (html) => {
    expect(html.find(`${mockComponentSelector('Working')} span`).text())
      .toMatch('Authenticating');
  },
  home: (html, ctx) => {
    expect(html.text()).toMatch('home');
    expect(ctx.mounted.find('Home').props()).toEqual(ctx.props);
  },
  error: (html) => {
    expect(html.find(mockComponentSelector('Error', 'headline')).text())
      .toMatch('Authentication error');
    const detail = html.find(mockComponentSelector('Error', 'detail'));
    expect(detail.text()).toMatch('Something went wrong');
    expect(detail.find(mockComponentSelector('RaisedButton', 'label')).text())
      .toMatch('Retry');
  },
};

const setup = ({ storage, fetch = () => {} }) => {
  const history = createHistory({ initialEntries: [homePath] });
  const store = configureStore(fromJS({}), history);
  const props = {
    prop1: 'test',
  };

  const tree = nest(
    [StoreProvider, { store }],
    [LanguageProvider, { messages }],
    [APIProvider, { base: api, fetch, middleware: [apiMiddleware] }],
    [AuthProvider, { storage, checkOn: LOCATION_CHANGE }],
    [RouterProvider, { history }],
    [ProtectedHome, props],
  );
  timestamp.mockReturnValue(date);
  return { tree, props };
};

const steps = [
  ...requestsSteps(api, requests),
  ...errorSteps(errors),
  ...reactSteps({ setup, screens }),
  ...storageSteps(datasets),
];


feature('containers/Auth/wrapper')
  .scenario('Rendering the wrapped component when the user is authenticated')

  .given('storage has current auth data')

  .when('the component tree is mounted')
  .then('I should see the home screen')
  .then('the rendered markup should match the snapshot')


  .scenario('Redirecting to sign-in when the user is not authenticated')

  .given('storage has empty auth data')

  .when('the component tree is mounted')
  .then('I should be redirected to the sign-in page')


  .scenario('Waiting for authentication, showing error and retrying')

  .given('storage has stale auth data')
  .given('refreshTokens request is expected')

  .when('the component tree is mounted')
  .then('I should see the placeholder screen')
  .then('the rendered markup should match the snapshot')
  .then('refreshTokens request should be made')

  .when('refreshTokens request fails with 500')
  .then('I should see the error screen')
  .then('the rendered markup should match the snapshot')

  .when('I click the retry button')
  .then('I should see the placeholder screen')

  .when('refreshTokens request fails with 401')
  .then('I should be redirected to the sign-in page')


  .step(/I should be redirected to the sign-in page/, (ctx) => {
    ctx.mounted.update();
    expect(ctx.mounted.find(mockComponentSelector('Redirect')).prop('to'))
      .toBe(`/signin?next=${encodeURIComponent(homePath)}`);
  })

  .step(/I click the retry button/, (ctx) => {
    ctx.mounted.update();
    ctx.mounted.find(mockComponentSelector('RaisedButton')).simulate('click');
  })

  .steps(steps)
  .run();
