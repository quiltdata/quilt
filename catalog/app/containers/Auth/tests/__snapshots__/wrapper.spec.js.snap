// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature: containers/Auth/wrapper Background: Scenario: Rendering the wrapped component when the user is authenticated Given storage has current auth data When the component tree is mounted Then the rendered markup should match the snapshot 1`] = `
<h1>
  home
</h1>
`;

exports[`Feature: containers/Auth/wrapper Background: Scenario: Waiting for authentication, showing error and retrying Given storage has stale auth data Given refreshTokens request is expected When the component tree is mounted Then the rendered markup should match the snapshot 1`] = `
<component-working>
  <span>
    Authenticating...
  </span>
</component-working>
`;

exports[`Feature: containers/Auth/wrapper Background: Scenario: Waiting for authentication, showing error and retrying Given storage has stale auth data Given refreshTokens request is expected When the component tree is mounted When refreshTokens request fails with 500 Then the rendered markup should match the snapshot 1`] = `
<component-error>
  <prop-headline>
    <span>
      Authentication error
    </span>
  </prop-headline>
  <prop-detail>
    <span>
      <span>
        Something went wrong. Try again.
      </span>
      <component-raised-button
        style="margin-left: 1em;"
      >
        <prop-label>
          <span>
            Retry
          </span>
        </prop-label>
      </component-raised-button>
    </span>
  </prop-detail>
</component-error>
`;
